flask_rebar/request_utils.py

```diff
@@ -18,6 +19,41 @@
 
 from flask_rebar import messages
 from flask_rebar import errors
+from flask_rebar.compat import MARSHMALLOW_MAJOR_VERSION
+
+
+if MARSHMALLOW_MAJOR_VERSION == 2:
+    def _normalize_result(result):
+        return result.data
+else:
+    def _normalize_result(result):
+        return result
+
+
+class HeadersProxy(Mapping):
```

Comment:
Doesn't this already exist in Python 3? https://docs.python.org/3/library/types.html#types.MappingProxyType (and for supporting python 2 as well there's https://pypi.org/project/dictproxyhack/)

Another one-line approach (which is also natively supported in Python 2) could be to call `Mapping.register(werkzeug.datastructures.Headers)`, making use of https://docs.python.org/3/library/abc.html#abc.ABCMeta.register
(Maybe that's a bit too close to monkey-patching / leaking side effects into users' applications. Hard to imagine that ever biting someone, but that's what they all say.)

Could also just wrap headers in a `dict()` call at every call site to keep things simpler (trading some additional (probably negligible) memory copying for fewer lines + simpler code).

Reply:
This is a really good thought, but there's a couple reasons I think why we should leave as is:

> Doesn't this already exist in Python 3? https://docs.python.org/3/library/types.html#types.MappingProxyType (and for supporting python 2 as well there's https://pypi.org/project/dictproxyhack/)

Unfortunately, EnvironHeaders.__iter__ yields tuples of (key, value). We need it to return keys to look like a `dict`. So this won't work: https://github.com/eevee/dictproxyhack/blob/master/dictproxyhack.py#L138

> Another one-line approach (which is also natively supported in Python 2) could be to call Mapping.register(werkzeug.datastructures.Headers), making use of https://docs.python.org/3/library/abc.html#abc.ABCMeta.register

I'd rather avoid mucking with this, for reasons you mention.

> Could also just wrap headers in a dict() call at every call site to keep things simpler (trading some additional (probably negligible) memory copying for fewer lines + simpler code).

Another good idea, but headers are case insensitive, which `EnvironHeaders` handles for us. Calling `dict` won'y quite cut it, since it will make the keys case sensitive.